### PR TITLE
[5.3] Fix artisan env error

### LINKED ISSUE #14941: Artisan Migrate env param has no effect
Essentially, running `artisan --env=testing migrate` does not actually choose the testing environment.

We have .env which is for development, but also .env.testing.  PHPUnit picks up the settings in .env.testing, but `artisan --env=testing migrate` does not.  It simply uses .env configuration.

I've seen some people solve this by adding a 'testing_\*' option in config/database.php, then using `artisan migrate --database=testing_*`, but this implies you need to always have a testing option in your database.php file, and seems like it shouldn't be that way.

The only way I got migrate to use .env.testing is to run `export APP_ENV=testing` prior to running the command.

Using laravel 5.3, php7.0, homestead 0.4.4 on vagrant box.

### PR DESCRIPTION
**Pull request by srmklive**
Fixes #14941. I was able to replicate the error reported, and came up with the fix in this PR. @GrahamCampbell @taylorotwell Kindly advise whether this PR's implementation is the right way to correct the reported issue.

### CODE REVIEW

File: src/Illuminate/Foundation/Console/Kernel.php [hunk 1]
```diff
@@ -294,6 +297,42 @@ protected function bootstrappers()
     }
 
     /**
+     * Set environment if 'env' option if provided.
+     *
+     * @return void
+     */
+    protected function getEnvironment()
+    {
+        foreach ($_SERVER['argv'] as $item) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: This is a security issue. You need to first verify that we're running in the console.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: That could be set by a web request and then someone calling the artisan facade.

Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: @GrahamCampbell What you suggest that i write here?

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: php_sapi_name === cli

Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: @GrahamCampbell updated. Is it correct now?

**Comment by srmklive** [comment]
@GrahamCampbell Will customize it further so its up to requirements.

File: src/Illuminate/Foundation/Console/Kernel.php [hunk 2]
```diff
@@ -294,6 +297,47 @@ protected function bootstrappers()
...
     /**
+     * Set environment if 'env' option if provided.
+     *
+     * @return void
+     */
+    protected function getEnvironment()
+    {
+        $app = app();
+
+        if ($app->runningInConsole()) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: Just do if it not, then early return

File: src/Illuminate/Foundation/Console/Kernel.php [hunk 3]
```diff
@@ -294,6 +297,47 @@ protected function bootstrappers()
     }
 
     /**
+     * Set environment if 'env' option if provided.
+     *
+     * @return void
+     */
+    protected function getEnvironment()
+    {
+        $app = app();
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: please don't do that

**Comment by srmklive** [comment]
PR updated with changes as per @GrahamCampbell's recommendation.

File: src/Illuminate/Foundation/Console/Kernel.php [hunk 4]
```diff
@@ -294,6 +299,33 @@ protected function bootstrappers()
...
 
     /**
+     * Set environment if 'env' option if provided.
+     *
+     * @param  InputInterface  $input
+     * @return void
+     */
+    protected function setEnvironment($input)
+    {
+        if (! app()->runningInConsole()) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: This function can never be used in the core. Please just call `$this->app->...`.

File: src/Illuminate/Foundation/Console/Kernel.php [hunk 5]
```diff
@@ -294,6 +299,33 @@ protected function bootstrappers()
...
+            return false;
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: just `return;` please

File: src/Illuminate/Foundation/Console/Kernel.php (hunk 4 above)
Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: @GrahamCampbell in this file `$this->app` implements the `Illuminate\Contracts\Foundation\Application` contract which doesn't have `environmentFile`, `loadEnvironmentFrom` functions. That's why i have used `app()` global helper here, so the `--env` option can be easily loaded.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: The app global helper is just the same. Please use $this->app.

Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: Yes indeed they are the same. If i use `$this->app` i can't load a different `.env` file, where as through app global helper i can.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: What?

Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: I mean to say if i use `$this->app`, i can't access the following methods:

* runningInConsole
* environmentFile
* environmentPath
* loadEnvironmentFrom

If i use `app`, i can access the above listed methods.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: yeh, you can. Laravel's contracts at stupid anyway. We use methods that arn't actually on them all over the place.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Console/Kernel.php]: We'll probably get rid of that application contract at some point anyway.

Comment (srmklive) [inline src/Illuminate/Foundation/Console/Kernel.php]: @GrahamCampbell If i write this line:

    $this->app->call(''Illuminate\Foundation\Application@method');

I can access the methods i mentioned without using app global helper.

**Comment by GrahamCampbell** [comment]
Please can you use `$this->app->` instead of the other nonsense. I've asked several times.

**Comment by taylorotwell** [comment]
There is no reason to use $this->app->call **every** time you call a method. That will be mega slow.

**Comment by taylorotwell** [comment]
This is also something that probably belongs in DetectEnvironment, not in the Console kernel.

**Pull request closed without merging**